File: LeetCode/Solutions/Python/21-MergeTwoSortedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if list1 is None:
            return list2
        elif list2 is None:
            return list1

        headNode = None

        currNode1 = list1
        currNode2 = list2

        if currNode1.val > currNode2.val:
            headNode = currNode2
            currNode2 = currNode2.next
        else:
            headNode = currNode1
            currNode1 = currNode1.next
        
        currNode = headNode

        while (currNode1 is not None) and (currNode2 is not None):
            if currNode1.val > currNode2.val:
                currNode.next = currNode2
                currNode = currNode2
                currNode2 = currNode2.next
            else:
                currNode.next = currNode1
                currNode = currNode1
                currNode1 = currNode1.next

        if currNode1 is None:
            currNode.next = currNode2
        elif currNode2 is None:
            currNode.next = currNode1

        return headNode

File: LeetCode/Solutions/Java/21-MergeTwoSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if(list1 == null) {
            return list2;
        } else if(list2 == null) {
            return list1;
        }

        ListNode headNode = null;
        ListNode currNode = headNode;

        ListNode list1Curr = list1;
        ListNode list2Curr = list2;

        // Assign the headnode, and currentnode
        if(list1Curr.val > list2Curr.val) {
            headNode = list2Curr;
            currNode = list2Curr;
            list2Curr = list2Curr.next;
        } else {
            headNode = list1Curr;
            currNode = list1Curr;
            list1Curr = list1Curr.next;
        }

        // Add nodes as long as either one is not empty
        while(list1Curr != null && list2Curr != null) {
            if(list1Curr.val > list2Curr.val) {
                currNode.next = list2Curr;
                currNode = list2Curr;
                list2Curr = list2Curr.next;
            } else {
                currNode.next = list1Curr;
                currNode = list1Curr;
                list1Curr = list1Curr.next;
            }
        }

        // When one of them ends first, link the other
        if(list1Curr == null) {
            currNode.next = list2Curr;
        } else if(list2Curr == null) {
            currNode.next = list1Curr;
        }

        return headNode;
    }
}